sw/host/opentitanlib/src/transport/proxy/mod.rs

```diff
@@ -0,0 +1,121 @@
+mod spi;
+mod uart;
+
+/// Implementation of the Transport trait backed by connection to a remost OpenTitan tool
+/// session process.
+pub struct Proxy {
+    inner: Rc<Inner>,
+}
+
+impl Proxy {
+    /// Establish connection with a running session process.
+    pub fn open(host: Option<&str>, port: u16) -> anyhow::Result<Self> {
+        let host = host.unwrap_or("localhost");
+        let addr = ToSocketAddrs::to_socket_addrs(&(host, port))
+            .map_err(|e| TransportError::ProxyLookupError(host.to_string(), e.to_string()))?
+            .next()
+            .unwrap();
+        let conn = TcpStream::connect(addr).map_err(|e| {
+            TransportError::ProxyConnectError(addr.to_string(), e.to_string())
+        })?;
+        Ok(Self {
+            inner: Rc::new(Inner {
+                reader: RefCell::new(BufReader::new(conn.try_clone()?)),
+                writer: RefCell::new(BufWriter::new(conn))
+            }),
+        })
+    }
+}
+
+struct Inner {
+    reader: RefCell<BufReader<TcpStream>>,
+    writer: RefCell<BufWriter<TcpStream>>,
+}
+
+impl Inner {
+    /// Helper method for sending one JSON request and receiving the response.  Called as part
+    /// of the implementation of every method of the sub-traits (gpio, uart, spi, i2c).
+    fn execute_command(&self, req: protocol::Request) -> Result<protocol::Response> {
+        self.send_json_request(req).wrap(TransportError::ProxyCommunicationError)?;
+        match self.recv_json_response().wrap(TransportError::ProxyCommunicationError)? {
+            Message::Res(res) => res,
+            _ => Err(TransportError::ProxyCommunicationError("Unexpected reply".to_string())),
+        }
+    }
+
+    /// Send a one-line JSON encoded requests, terminated with one newline.
+    fn send_json_request(&self, req: protocol::Request) -> anyhow::Result<()> {
+        let mut conn = self.writer.borrow_mut();
+        serde_json::to_writer(&mut *conn, &Message::Req(req))?;
+        conn.write(&[b'\n'])?;
+        conn.flush()?;
+        Ok(())
+    }
+
+    /// Receive until newline, and decode one JSON response.
+    fn recv_json_response(&self) -> anyhow::Result<Message> {
+        let mut conn = self.reader.borrow_mut();
+        let mut buf = Vec::new();
+        let mut idx: usize = 0;
+        loop {
+            buf.resize(idx + 2048, 0);
+            let rc = conn.read(&mut buf[idx..])?;
+            if rc == 0 {
+                anyhow::bail!(io::Error::new(ErrorKind::UnexpectedEof, "Truncated JSON"))
+            }
+            idx += rc;
+            if buf[idx - 1] == b'\n' {
+                break;
+            }
+        }
+        Ok(serde_json::from_slice::<Message>(&buf[..idx - 1])?)
+    }
+}
+
+impl Transport for Proxy {
+    fn capabilities(&self) -> Capabilities {
+        Capabilities::new(Capability::UART | Capability::GPIO | Capability::SPI | Capability::I2C)
+    }
+
+    // Crate SPI Target instance, or return one from a cache of previously created instances.
```

Comment:
nit: s/Crate/create ? applies throughout

Reply:
Argh.  9 out of 10 times when I need to type "use crate::...", I end up typing "create" instead.  It seems that my fingers cannot even get one of the two right consistently.